packages/openid4vc/src/openid4vc-issuer/router/redirectEndpoint.ts

```diff
@@ -0,0 +1,190 @@
+
+        if (!chainedIdentityState || typeof chainedIdentityState !== 'string') {
+          throw new Oauth2ServerErrorResponseError({
+            // Server error because it's an error of the external IDP
+            error: Oauth2ErrorCodes.ServerError,
+            error_description: `Missing required 'state' parameter`,
+          })
+        }
+
+        issuanceSession = await openId4VcIssuerService.findSingleIssuanceSessionByQuery(agentContext, {
+          issuerId: issuer.issuerId,
+          chainedIdentityState: chainedIdentityState,
+        })
+
+        if (!issuanceSession || issuanceSession.state !== OpenId4VcIssuanceSessionState.AuthorizationInitiated) {
+          throw new Oauth2ServerErrorResponseError(
+            {
+              error: Oauth2ErrorCodes.InvalidRequest,
+              error_description: `Invalid 'state' parameter`,
+            },
+            {
+              internalMessage: !issuanceSession
+                ? `Issuance session not found for identity chaining 'state' parameter '${chainedIdentityState}'`
+                : `Issuance session '${issuanceSession.id}' has state '${
+                    issuanceSession.state
+                  }' but expected ${OpenId4VcIssuanceSessionState.AuthorizationInitiated}`,
+            }
+          )
+        }
+
+        if (
+          !issuanceSession.chainedIdentity?.externalAuthorizationServerUrl ||
+          !issuanceSession.chainedIdentity.redirectUri
+        ) {
+          throw new Oauth2ServerErrorResponseError(
+            {
+              error: Oauth2ErrorCodes.InvalidRequest,
+              error_description: 'The session is invalid or has expired.',
+            },
+            {
+              internalMessage: `Issuance session '${issuanceSession.id}' does not have identity chaining configured, so it's not compatible with the redirect endpoint.`,
+            }
+          )
+        }
+
+        if (chainedIdentityAuthorizationCode && typeof chainedIdentityAuthorizationCode === 'string') {
+          // TODO: is storing the whole metadata a good idea? Or should we refetch? What if it has changed?
+          const authorizationServerMetadata = issuanceSession.chainedIdentity.externalAuthorizationServerMetadata
+          if (!authorizationServerMetadata) {
+            throw new Oauth2ServerErrorResponseError(
+              {
+                error: Oauth2ErrorCodes.ServerError,
+                error_description: `Unable to retrieve authorization server metadata from external identity provider.`,
+              },
+              {
+                internalMessage: `Missing chained identity server metadata on issuance session '${issuanceSession.id}'`,
+              }
+            )
+          }
+
+          const oauth2Client = openId4VcIssuerService.getOauth2Client(agentContext, issuer)
+
+          // Retrieve / verify access token.
+          // TODO: add support for DPoP
+          const { accessTokenResponse } = await oauth2Client
```

Comment:
See earlier comment about needing callback, or also supporting the specific response type as used by e.g. google.

If the `openid` scope is included, there will be an id token in the access token response as well. We need to verify this token. I think verifying an ID Token is quite simple, so we might want to just do that. But open to other solutions as well.

Or we should explicitly clear that the id token is not verified (by e.g. placing it in a separate `unverifiedAccessTokenResponse`).